iroh-bytes/src/store/traits.rs

```diff
@@ -103,23 +110,178 @@ pub trait Map: Clone + Send + Sync + 'static {
+    FallibleProgressBatchWriter<W, F>
+{
+    /// Create a new `FallibleProgressBatchWriter` from an inner writer and a progress callback
+    ///
+    /// The `on_write` function is called for each write, with the `offset` as the first and the
+    /// length of the data as the second param. `on_write` must return an `io::Result`.
+    /// If `on_write` returns an error, the download is aborted.
+    pub fn new(inner: W, on_write: F) -> Self {
+        Self(inner, on_write)
+    }
+
+    /// Return the inner writer.
+    pub fn into_inner(self) -> W {
+        self.0
+    }
+}
+
+impl<W: BaoBatchWriter, F: Fn(u64, usize) -> io::Result<()> + 'static> BaoBatchWriter
+    for FallibleProgressBatchWriter<W, F>
+{
+    fn write_batch(
+        &mut self,
+        size: u64,
+        batch: Vec<BaoContentItem>,
+    ) -> LocalBoxFuture<'_, io::Result<()>> {
+        // todo: remove all this boxing once we got 1.75
+        async move {
+            // find the offset and length of the first (usually only) chunk
+            let chunk = batch
+                .iter()
+                .filter_map(|item| {
+                    if let BaoContentItem::Leaf(leaf) = item {
+                        Some((leaf.offset.0, leaf.data.len()))
+                    } else {
+                        None
+                    }
+                })
+                .next();
+            self.0.write_batch(size, batch).await?;
+            // call the progress callback
+            if let Some((offset, len)) = chunk {
+                (self.1)(offset, len)?;
+            }
+            Ok(())
+        }
+        .boxed_local()
+    }
+
+    fn sync(&mut self) -> LocalBoxFuture<'_, io::Result<()>> {
+        self.0.sync()
+    }
+}
+
+/// A combined batch writer
+///
+/// This is just temporary to allow reusing the existing store implementations
+/// that have separate data and outboard writers.
+#[derive(Debug)]
+pub struct CombinedBatchWriter<D, O> {
+    /// data part
+    pub data: D,
+    /// outboard part
+    pub outboard: O,
+}
+
+impl<D, O> BaoBatchWriter for CombinedBatchWriter<D, O>
+where
+    D: AsyncSliceWriter,
+    O: OutboardMut,
+{
+    fn write_batch(
+        &mut self,
+        _size: u64,
+        batch: Vec<BaoContentItem>,
+    ) -> LocalBoxFuture<'_, io::Result<()>> {
+        async move {
+            for item in batch {
+                match item {
+                    BaoContentItem::Parent(parent) => {
+                        self.outboard.save(parent.node, &parent.pair).await?;
+                    }
+                    BaoContentItem::Leaf(leaf) => {
+                        self.data.write_bytes_at(leaf.offset.0, leaf.data).await?;
+                    }
+                }
+            }
+            Ok(())
+        }
+        .boxed_local()
+    }
+
+    fn sync(&mut self) -> LocalBoxFuture<'_, io::Result<()>> {
+        async move {
+            self.data.sync().await?;
+            self.outboard.sync().await?;
```

Comment:
could these be polled using `join`?

Reply:
yeah, but this thing is temporary anyway. The idea of the batch writer in the long term is to be able to batch things including sync, and to leave it up to the impl what needs to be synced.

So the long term impl would be spawn_blocking(|| { sync all the files }) or something.